ext/drstatecmp/drstatecmp.c

```diff
@@ -0,0 +1,572 @@
+    drstatecmp_check_gpr_value("r19", mc_instrumented->r19, mc_expected.r19);
+    drstatecmp_check_gpr_value("r20", mc_instrumented->r20, mc_expected.r20);
+    drstatecmp_check_gpr_value("r21", mc_instrumented->r21, mc_expected.r21);
+    drstatecmp_check_gpr_value("r22", mc_instrumented->r22, mc_expected.r22);
+    drstatecmp_check_gpr_value("r23", mc_instrumented->r23, mc_expected.r23);
+    drstatecmp_check_gpr_value("r24", mc_instrumented->r24, mc_expected.r24);
+    drstatecmp_check_gpr_value("r25", mc_instrumented->r25, mc_expected.r25);
+    drstatecmp_check_gpr_value("r26", mc_instrumented->r26, mc_expected.r26);
+    drstatecmp_check_gpr_value("r27", mc_instrumented->r27, mc_expected.r27);
+    drstatecmp_check_gpr_value("r28", mc_instrumented->r28, mc_expected.r28);
+    drstatecmp_check_gpr_value("r29", mc_instrumented->r29, mc_expected.r29);
+#    endif
+
+    drstatecmp_check_gpr_value("lr", mc_instrumented->lr, mc_expected.lr);
+    drstatecmp_check_xflags_value("xflags", mc_instrumented->xflags, mc_expected.xflags);
+
+#else
+#    error NYI
+#endif
+
+    drstatecmp_check_gpr_value("xsp", mc_instrumented->xsp, mc_expected.xsp);
+    for (int i = 0; i < MCXT_NUM_SIMD_SLOTS; i++) {
+        drstatecmp_check_simd_value(&mc_instrumented->simd[i], &mc_expected.simd[i]);
+    }
+}
+
+static void
+drstatecmp_compare_state(void *drcontext, instrlist_t *bb, instr_t *instr)
+{
+    dr_insert_clean_call(drcontext, bb, instr, (void *)drstatecmp_compare_state_call,
+                         false /*fpstate */, 0);
+}
+
+static void
+drstatecmp_insert_instrument_bb_with_side_effects(void)
+{
+    /* TODO i#4678: Add checks for bbs with side effects.  */
+}
+
+static void
+drstatecmp_insert_instrument_bb_orig_side_effect_free(void *drcontext, instrlist_t *bb,
+                                                      instr_t *instr,
+                                                      drstatecmp_user_data_t *data)
+{
+    /* Save state at the beginning of this bb in order to restore it at the end of
+     * the bb (to enable re-execution of the bb). */
+    if (match_label_val(instr, DRSTATECMP_LABEL_ORIG_BB)) {
+        drstatecmp_save_state(drcontext, bb, instr, /*for_cmp=*/false);
+    }
+
+    /* Change mode to process the copy bb once we encounter the COPY_BB label.
+     * Save the state at the end of this bb for later comparison, and remove the
+     * previously inserted jmp instr to the TERM label of the copy bb and let it fall
+     * through to the instructions of the copy bb (that will be instrumentation-free).
+     * Finally, restore the machine state to the state before executing the original
+     * version of this bb (allows us to re-execute this bb).
+     */
+    else if (match_label_val(instr, DRSTATECMP_LABEL_COPY_BB)) {
+        data->orig_bb_mode = false;
+        instr_t *prev_instr = instr_get_prev_app(instr);
+        instrlist_remove(bb, prev_instr);
```

Comment:
If I am not mistaken, you cannot remove app instructions (inserted at app2app) during the insertion phase. This breaks DR's contract.

Moreover, there could have been an application code analysis pass (so between app2app and insertion) that takes into account the jmp.

Reply:
Addressed by moving all changes to the new final phase.